Fix wrong optimisation of arguments[-1]

rhuanjl opened the pull request:
**Background**
The use of the full `arguments` object per spec in Javascript has various performance penalties. For this reason the CC Jit contains two different optimisations for it which seek to optimise away the use of this object when it is being used for simple purposes that can be redirected (e.g. accessing params that exist)

One of these optimisations is specific to inlined functions and the other is slightly more general, the inline specific optimisation was added in the master branch since the last release and introduced a bug.

**Specific Bug**
`arguments[-1]` in function inlined by the Jit was loading the `this` object - when it should return undefined (and would return undefined when not inlined).

This is because Chakracore internally stores arguments to JS functions in a row with the `this` object as item 0 and a check was not being done to avoid hitting it.

**Ride-along improvement**
The optimisation of `arguments` in a non-inlined function was triggering a rejit if it encountered a constant negative index - but such should just result in returning undefined - added a path for doing that, see comments below.

**EDIT:** converted to draft whilst I look at the issue @pleath has raised below.

FIx #6783

rhuanjl commented on lib/Backend/GlobOpt.cpp:
```diff
@@ -13525,16 +13526,19 @@ GlobOpt::OptStackArgLenAndConst(IR::Instr* instr, Value** src1Val)
-                        }
-                        return false;
-                    });
+                    if (argIndex > 0)
+                    {
+                        IR::Instr* inlineeStart = instr->m_func->GetInlineeStart();
+                        inlineeStart->IterateArgInstrs([&](IR::Instr* argInstr) {
+                            StackSym *argSym = argInstr->GetDst()->AsSymOpnd()->m_sym->AsStackSym();
+                            if (argSym->GetArgSlotNum() - 1 == argIndex)
+                            {
+                                defInstr = argInstr;
+                                return true;
+                            }
+                            return false;
+                        });
+                    }
```
This is the bugFix, in the case that `arguments[-1]` was used, argIndex would be 0 (see line 13527) but there would be an argSym with slot value 0 - namely the `this` object. Avoid incorrectly matching it by not searching for a match unless the index is > 0.

`arguments[-n]` for any `n` should return `undefined` - which is what we produce below if `defInstr` is left as `nullptr`

pleath replied:
So you say this is assuming that arguments[-n] is always "undefined"? That doesn't sound right.

+function() {
    arguments[-1] = "it's me!";
    WScript.Echo(arguments[-1]);
}();

rhuanjl replied:
A very good point, I need to look at how this case is handled - and add a test for it. I've converted the PR to draft for now as this may make it significantly more complicated.

pleath replied:
I think this should be fairly simple. As with arrays, a load of a[n] where n is not in 0...a.length-1 should just be forwarded to the load-element runtime helper. That case isn't important to optimize anyway, and the semantics around array gaps, prototype objects, etc., aren't something we should try to represent in jitted code.

pleath replied:
BTW, it's weird to me that this actually passed the unit tests. We're obviously missing some coverage for this kind of case.

rhuanjl replied:
There is certainly missing coverage.

But also currently the optimisation is disabled in advance if the arguments object is written to - so for any out of range numerically indexed property field the correct value within the optimisation is undefined.

I think the only currently detectable incorrect behaviour is with arguments[-1] specifically in an inlined function picking up the this value.

My attempt to remove a few rejits may add another incorrect case though I'd still like to double check that.

rhuanjl commented on lib/Backend/Lower.cpp:
```diff
@@ -22183,72 +22184,81 @@ Lowerer::GenerateFastArgumentsLdElemI(IR::Instr* ldElem, IR::LabelInstr *labelFa
         {
-            argIndirOpnd = GetArgsIndirOpndForTopFunction(ldElem, valueOpnd);
+            IR::Opnd *undef = LoadLibraryValueOpnd(ldElem, LibraryValue::ValueUndefined);
+            Lowerer::InsertMove(ldElem->GetDst(), undef, ldElem);
```
This logic is the ride-along for functions that are not (yet) inlined, when attempting to jit `arguments[-n]` before it would always fail to emit the fastpath here which triggered a rejit. Instead now we insert a fast load of undefined.

(Whilst `arguments[-n]` is not the sort of logic we expect in any nice JS it seems silly to force a rejit for it when the per-spec action is simply to load undefined)

pleath replied:
How does this work if the correct value of arguments[-1] is other than undefined?

rhuanjl committed:
Fix Optimisation of arguments[-n]

rhuanjl commented:
@ppenzin I've added tests for the points discussed with pleath above.

Upon looking at the logic further I'm happy that my actual code was fine though.

There are two different optimisations here - for loads of `arguments[x]` one for a normal jitted function and the other for an inlined function. In both cases the optimisation is disabled earlier in the Jit process if the arguments object is edited in any way - hence why it is safe to supply undefined for any out of bounds load.

EDIT: test fail needs investigating - may be a legitimate issue I've somehow introduced or may need a new baseline;

ppenzin commented on test/Optimizer/StackArgumentsOptNegativeIndex.js:
```diff
@@ -0,0 +1,99 @@
+    return arguments[-1];
+}
+
+for (i = 0; i < 500; i++) {
```
Are the loops added to ensure JITing?

rhuanjl replied:
To get profile info and then jit - may not be needed if I set it up to work correctly with the dynopogo test setup but this just did the job simply.

ppenzin reviewed (approved):
This looks good to me. If test failures are just wrong baselines, you can just merge it in.

rhuanjl commented:
I've updated a few comments for clarity and I believe I've solved the test issues.

Test 1: "funcexpr_2"
My ride along improvement had got rid of a rejit in this testcase which in turn meant the baseline was wrong. The case is meant to test a facility used in that rejit so I edited the case so the rejit still occurs.

Test 2: "StackArgWIthFormals"
My ride along improvement had got rid of a bailout in this testcase which meant the baseline was wrong. As other cases in this testfile hit similar bailouts I've updated the baseline.
